Describe the use cases for events in greater detail (#3969)
open-telemetry/opentelemetry-specification: specification/logs/event-api.md

* [EventLoggerProvider](#eventloggerprovider) is the entry point of the API. It
  provides access to `EventLogger`s.
* [EventLogger](#eventlogger) is the component responsible for emitting events.

## Event Data model

Wikipedia’s [definition of log file](https://en.wikipedia.org/wiki/Log_file):

>In computing, a log file is a file that records either events that occur in an
>operating system or other software runs.

From OpenTelemetry's perspective LogRecords and Events are both represented
using the same [data model](./data-model.md). An Event is a specialized type
of LogRecord, not a separate concept.

Events are OpenTelemetry's standardized semantic formatting for LogRecords.
Beyond the structure provided by the LogRecord data model, it is helpful for
logs to have a common format within that structure. When OpenTelemetry
instrumentation emits logs, those logs SHOULD be formatted as Events. All
semantic conventions defined for logs MUST be formatted as Events.

The Event format is as follows. All Events have a `name` attribute, and all
Events with the same `name` MUST conform to the same schema for both their
`Attributes` and their `Body`.

> ## Event API usecases

Reviewer: ```suggestion
## Event API use cases
```

The Events API was designed to allow shared libraries to emit high quality
logs without needing to depend on a third party logger. Unlike the
[Logs Bridge API](./bridge-api.md), instrumentation authors and application
developers are encouraged to call this API directly. It is appropriate to
use the Event API when these properties fit your requirements:

* Logging from a shared library that must run in many applications.
* A semantic convention needs to be defined. We do not define semantic
  conventions for LogRecords that are not Events.
* Analysis by an observability platform is the intended use case. For
  example: statistics, indexing, machine learning, session replay.
* Normalizing logging and having a consistent schema across a large
  application is helpful.

If any of these properties fit your requirements, we recommend using the Event API.
Events are described in more detail in the [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md).

> Please note that Events are sent directly to the OTel Log SDK, which currently

Reviewer: There may be other implementations of Events API then the SDK.

They may custom Event API implementations which would support it.

I think we should not couple API spec to the SDK.

Author: I clarified in this section that these limitations only relate to the OTel SDK implementation. 

I would lean in to your suggestion to move things to the README and say that both the data model and use case sections should be moved somewhere else, either the README, the log data model doc, or a new event data model doc. But I'm hesitant to do that in this PR, it feels like opening a can of worms and for now I would just like to record the relevant information for this experimental API in the same place we have been recording it so far, as the purpose of this PR is to update the content of these sections.

If someone wants to make a follow up PR to move this information, I suggest a `event-data-model` doc would be the best place to put it.

Reviewer: I am fine with your changes and I approved the PR 😉

> features present in popular log frameworks. For example: pattern logging, file

Reviewer: ditto

```suggestion
features present in popular logging libraries. For example: pattern logging, file
```

pattern logging, file rotation, network appenders, etc. These features cannot be
used with the Event API at this time.

> If a logging framework is capable of creating logs which correctly map

Reviewer: For consistency with e.g. https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/logs

```suggestion
If a logging library is capable of creating logs which correctly map
```

>operating system or other software runs.

From OpenTelemetry's perspective LogRecords and Events are both represented
using the same [data model](./data-model.md). An Event is a specialized type
of LogRecord, not a separate concept.

Events contain all of the features provided by LogRecords, plus one additional
feature. All Events have a `name`.  Events with the same `name` MUST conform to
the same schema for both their `Attributes` and their `Body`.

Unlike the [Logs Bridge API](./bridge-api.md), application developers and
instrumentation authors are encouraged to call this API directly. It is
appropriate to use the Event API when these properties fit your requirements:

* A consistent schema that can be identified by a name is necessary.
* Analysis by an Observability platform is the intended use case. For
  example: statistics, indexing, machine learning, session replay.
* A semantic convention needs to be defined. We do not define semantic
  conventions for LogRecords that are not Events.

> If any of these properties fit your requirements, we recommend using the Event API.

Reviewer 2: Would this apply to the programming languages / runtimes that already have established logging API and the logging API already supports event name?

Author: I would say yes. It is a recommendation, not a requirement. Going through a different API risks a translation error. If the intent is to send OpenTelemetry an Event, we recommend that you use the API we provide.

Author: I added a note lower down where we talk about other log frameworks, saying that it is acceptable to use them as long as it actually works.

>operating system or other software runs.

From OpenTelemetry's perspective LogRecords and Events are both represented
using the same [data model](./data-model.md). An Event is a specialized type
of LogRecord, not a separate concept.

Events contain all of the features provided by LogRecords, plus one additional

> feature. All Events have a `name`.  Events with the same `name` MUST conform to
> the same schema for both their `Attributes` and their `Body`.

Reviewer: > Events with the same `name` MUST conform to the same schema for both their `Attributes` and their `Body`.

Two whom this requirement is addressed?

Event API users? If so then we cannot control how they are using it. Using `MUST` per [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) is not really applicable.

I guess that it is about semantic conventions, but it is not clear

```suggestion
feature. All Events have a `Name`. The semantic conventions MUST
define events with the same `Name` to conform to the same schema
for both their `Attributes` and their `Body`.
```

PS. For me this sentence is hard to understand. I guess that it means that for given schema URL, events with given `Name` for share the same schema (`Attributes` and `Body` structure and semantics).

Author: To clarify, using MUST in this context is valid. There are many requirement in OpenTelemetry which could be circumvented by the end user, including everything related to semantics. I'll update this language to make it clearer that every event is a semantic convention, I agree that is getting lost here.

Unlike the [Logs Bridge API](./bridge-api.md), application developers and
instrumentation authors are encouraged to call this API directly. It is
appropriate to use the Event API when these properties fit your requirements:

* A consistent schema that can be identified by a name is necessary.

> * Analysis by an Observability platform is the intended use case. For

Reviewer 2: ```suggestion
* Analysis by an observability platform is the intended use case. For
```

  example: statistics, indexing, machine learning, session replay.
* A semantic convention needs to be defined. We do not define semantic
  conventions for LogRecords that are not Events.

If any of these properties fit your requirements, we recommend using the Event API.
Events are described in more detail in the [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md).

Please note that Events are sent directly to the OTel Log SDK, which currently lacks a
number of advanced features present in popular log frameworks. For example:
pattern logging, file rotation, network appenders, etc. These features cannot be

> used with the Event API at this time. If a log framework is capable of creating
> logs which correctly map to the Event data model, that is also an acceptable
> way to create Events.

Reviewer 2: Maybe make this stronger:

```suggestion
used with the Event API at this time. For each programming language/runtime, if an established logging framework is capable of creating
logs which correctly map to the Event data model, it is RECOMMENDED to promote the logging framework rather than inventing a competing solution.
```

Author: I disagree. What competing solution is being invented? We are not intending to suggest that end users avoid the Event API, we are just clarify that since events are logs, they can also be created using a logger.

Author: I updated this section to clarify when we recommend using a third party logger, and when we recommend using the Event API directly.

to call this API directly. It is appropriate to use the Event API when these properties fit your requirements:

* Logging from a shared library that must run in many applications.
* A semantic convention needs to be defined. We do not define semantic
  conventions for LogRecords that are not Events.
* Analysis by an observability platform is the intended use case. For
  example: statistics, indexing, machine learning, session replay.
* Normalizing logging and having a consistent schema across a large
  application is helpful.

If any of these properties fit your requirements, we recommend using the Event API.
Events are described in more detail in the [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md).

Please note that Events are sent directly to the OTel Log SDK, which currently
lacks a number of advanced features present in popular log frameworks. For example:
pattern logging, file rotation, network appenders, etc. These features cannot be
used with the Event API at this time.

If a third party logging framework is capable of creating logs which correctly map

> to the Event data model, loggin in this manner is also an acceptable way to create

Reviewer 2: ```suggestion
to the Event data model, logging in this manner is also an acceptable way to create
```

Events. For application developers that need additional logging features, we
recommend using this approach. For shared libraries and instrumentation, we recommend

> using the Event API directly, to avoid taking a dependecy on a third party logger.

Reviewer 2: What if this is not a 3rd party logger? For example, it is part of the language runtime or it is part of a framework?

Author: That is still forcing the user to figure out which logger every library is using.

Reviewer 2: Would you elaborate a bit?

Here is an example how folks might handle it in the .NET ecosystem:
1. Each library is logging using an abstraction provided by the .NET runtime libraries. Libraries can decide to request for a single logger, or multiple instances of logger in case they want to separate the concern or provide certain isolation.
2. The actual implementation is configured by the application developer.
3. The dependency injection engine (which is part of the runtime libraries) will connect 1 and 2, so the user configured loggers can be passed to the libraries, or the user configured logger factory instances can be passed to the libraries then the libraries can use the factory instances to create dedicated loggers.

Author: .NET is an extreme outlier. Most languages do not work like this. We don't write the spec with .NET's uniqueness in mind, it diverges from the spec in many ways.

Author: We don't write in our tracing API "do not use the OpenTelemetry tracing API, we recommend using a language or framework tracing API instead."

Author: btw, you can find the more in-depth discussion on the use of the Event API in #4045, the "shared OpenTelemetry instrumentation" section describes the problem in more detail. It is great that .NET has addressed this exact problem, but unfortunately that is not the case in any other language. Log4J is the only other thing out there that comes close.

Reviewer 2: > Log4J is the only other thing out there that comes close.

Based on what I've observed, I feel Rust is probably going to be another one.

Reviewer 2: > We don't write in our tracing spec "do not use the OpenTelemetry Tracing API, we recommend using a language or framework tracing API instead."

I think for logging we've decided to take a different approach https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/logs#introduction.

* _For metrics and traces OpenTelemetry takes the approach of a clean-sheet design, specifies a new API and provides full implementations of this API in multiple languages._
* _Our approach with logs is somewhat different. For OpenTelemetry to be successful in logging space we need to support existing legacy of logs and logging libraries, while offering improvements and better integration with the rest of observability world where possible._

My question point is not specific about .NET at all, I was just using .NET as an example.

Think about this - if there is already a widely used and well-established logging API for certain programming language, and that logging API is capable of supporting the "events" scenario, do we still want OpenTelemetry language SIGs to create one more? If a particular language runtime already has logging API, and the owners of the logging API are committed to evolving such API to be fully compatible with OpenTelemetry specification, would we want to collaborate with them so OpenTelemetry APIs can be embeded in the runtime (so that the runtime itself can use OpenTelemetry API to instrument runtime components)?

Author: Again, if there is a special case in a particular language, we handle that as a special case. 

With logging, somehow the script has been flipped? Because there might be some special cases, we should not provide a solution for the majority of cases?

I agree that rust is a special case for tracing. Does that mean we should worry about having a tracing api in the spec?

Reviewer 2: > I agree that rust is a special case for tracing. Does that mean we should worry about having a tracing api in the spec?

No. I think what we can probably do is to choose one of the following options:

1. Keep the tracing spec as-is, there is probably nothing to be updated. The OpenTelemetry Rust community can tell the user "OpenTelemetry Rust project itself doesn't ship tracing API, instead, the OTel Rust and Tokio communities work closely to build tracing API in Tokio-tracing, and users should use that".
2. Update the tracing spec by providing some guidance, such as "we encourage language runtime and framework owners to collaborate with our community, adopt OpenTelemetry standard and build the spec-compliant OpenTelemetry Tracing APIs in the framework/runtime to make deeper/bigger impact".

Reviewer 2: > Again, if there is a special case in a particular language, we handle that as a special case.

Hmm, I actually think this is a general question regarding the direction, rather than a question about specific cases. And that's why I want us to think deeper and have a clear vision/guidance.

Author: Ok, well there is a detailed explanation for this reasoning in #4045, I would encourage you to review it as it does contain this clear vision and guidance, based on a lot of deep thinking as we have developed events.

Traditionally we do not include this kind of reasoning in the spec. Maybe we should, but I see no reason to treat the Event API differently from how we write down our other APIs in the spec. The logging SIG never intended to preclude us from building something like the Event API if we needed it, we just did not want to jump the gun and create some kind of generic logging system like the ones already in existence.

To summarize the current state of logs and events in OpenTelemetry, this is my understanding:

* Applications already have logging. Consuming those logs without forcing users to reinstrument is the goal of the OpenTelemetry Log Bridge API.
* OpenTelemetry has no need for a general purpose logging system in any language, with string parsing and all the rest of the features those logging systems have. So we have no intention of building one of these.
* OpenTelemetry does intend to produce structured logs to describe computer operations. Our event model defines how we intend to structure these logs.
* OpenTelemetry instrumentation needs to emit these events, so we have an Event API. Read #4045 if you want an explanation for why 3rd party loggers do not solve this use case for us.
* OpenTelemetry believes that events are the best format for creating logs, which is why we chose it as our design. Of course we encourage end users to log using the event format, and to use our Event API if they want to. There is no need to confuse users by creating caveats for this.
* If for some reason end users want to send events recorded via the Event API back through their logging system, we would like to accommodate them. We haven't finished the design for this yet, so for the time being we have a warning in the spec saying that this does not work yet.

For the above reasons, I feel that we should keep our description of the Event API simple in the spec, and not treat it differently from how we describe our other APIs. Is that helpful?

Reviewer 2: > For the above reasons, I feel that we should keep our description of the Event API simple in the spec, and not treat it differently from how we describe our other APIs. Is that helpful?

Thanks! I agree with "keep our description of the Event API simple in the spec".

In this case I suggest changing the wording from:

```text
If a third party logging framework is capable of creating logs which correctly map
to the Event data model, logging in this manner is also an acceptable way to create
Events. For application developers that need additional logging features, we
recommend using this approach. For shared libraries and instrumentation, we recommend
using the Event API directly, to avoid taking a dependency on a third party logger.
```

to:

```
If a logging framework is capable of creating logs which correctly map
to the Event data model, logging in this manner is also an acceptable way to create
Events.
```

Reviewer 3: > Ok, well there is a detailed explanation for this reasoning in #4045, I would encourage you to review it as it does contain this clear vision and guidance, based on a lot of deep thinking as we have developed events.

my preference would be to record that information somewhere other than the issue. If we previously had an OTEP for events, we can update it and include these motivations. Or create a new OTEP. Then, as was recently discussed, we can reference the OTEP(s) from the Spec to avoid repeating the reasoning in the spec.

Author: Thanks @reyang, I agree. I've updated the PR with your suggestion.

@yurishkuro that makes sense, I will update the events OTEP with this information.

>operating system or other software runs.

From OpenTelemetry's perspective LogRecords and Events are both represented
using the same [data model](./data-model.md). An Event is a specialized type
of LogRecord, not a seperate concept.

Events contain all of the features provided by LogRecords, plus one additional
feature. All Events have a `name`.  Events with the same `name` MUST conform to
the same schema for both their `Attributes` and their `Body`.

Unlike the [Logs Bridge API](./bridge-api.md), application developers and
instrumentation authors are encouraged to call this API directly. It is 
appropriate to use the Event API when these properties fit your requirements:

* A consistent schema that can be identified by a name is necessary.
* A semantic convention needs to be defined. We do not define semantic
  conventions for logs.
* Analysis by an Observability platform is the intended use case. For
  example: statistics, indexing, machine learning, session replay.

If any of these properties fit your requirements, we recommend using the Event API.

> Events are described in more detail in the[semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md).

Reviewer 4: space missing? (`the [semantic`)

>operating system or other software runs.

From OpenTelemetry's perspective LogRecords and Events are both represented
using the same [data model](./data-model.md). An Event is a specialized type
of LogRecord, not a separate concept.

Events contain all of the features provided by LogRecords, plus one additional

> feature: every event is a semantic convention. All Events have a

Author: The intention is to say that the Event format is how we define semantic conventions for LogRecords. OTel instrumentation should use the Event format, and not create logs with arbitrary structures. Maybe there is a better way to describe this.

> feature: every event is a semantic convention. All Events have a

Author: I've changed this section to clarify that Events are just a format for semantic conventions that all OpenTelemetry instrumentation should use when emitting logs.

`name`, and all Events with the same `name` MUST conform to the same schema for

> both their `Attributes` and their `Body`. This additional 

Reviewer 5: ```suggestion
both their `Attributes` and their `Body`.
```

The Events API was designed to allow shared libraries to emit high quality
logs without needing to depend on a third party logger. Unlike the
[Logs Bridge API](./bridge-api.md), instrumentation authors and application

> developers and are encouraged to call this API directly. It is appropriate to

Reviewer 6: ```suggestion
developers are encouraged to call this API directly. It is appropriate to
```
